File: telegram-service/src/main/java/de/nak/home_assistant/actions/ModuleToggle.java
package de.nak.home_assistant.actions;

import com.pengrad.telegrambot.request.SendMessage;
import de.nak.home_assistant.CustomKeyboard;
import de.nak.home_assistant.models.service.AuthResponse;
import de.nak.home_assistant.models.service.ServiceRequest;
import de.nak.home_assistant.models.telegram.CallbackData;
import de.nak.home_assistant.services.ServiceService;
import de.nak.home_assistant.services.database.UserService;

public class ModuleToggle {

    public static SendMessage generateMessage(long userId, long chatId, CallbackData cbData, boolean activate) {
        int moduleId = cbData.getModuleId();
        UserService userService = new UserService(userId);

        String text;
        if (activate) {
            ServiceRequest serviceRequest = userService.activateModule(moduleId);
            ServiceService serviceService = new ServiceService(new AuthResponse());
            AuthResponse authResponse = serviceService.makeRequest(serviceRequest);

            if (authResponse.isSuccess()) {
                text = authResponse.getResult();
            } else {
                text = "Etwas ist schiefgegangen " + authResponse.getError();
            }
        } else {
            if (userService.deactivateModule(moduleId)) {
                text = "Modul deaktiviert";
            } else {
                text = "Etwas ist schiefgegangen";
            }
        }

        return CustomKeyboard.generateMessageWithKeyboard(userId, chatId, text);
    }
}

File: packages/telegram-service/src/actions/module-toggle.ts
import { AuthResponse, CallBackData } from "@microhome/types";
import { bot } from "../bot";
import { i18n } from "../i18n";
import { dashboardKeyboardMarkup } from "../keyboard";
import { makeServiceRequest } from "../services/service";
import { activateModule, deactivateModule } from "../services/user";

export const sendModuleToggle = async (
  userId: number,
  cbData: CallBackData,
  activate: boolean
) => {
  const translations = i18n();

  const { moduleId } = cbData;

  if (!moduleId) {
    return;
  }

  let text;

  if (activate) {
    const serviceRequest = await activateModule(userId, moduleId);

    if (!serviceRequest) {
      return;
    }

    const { error, success, result } = await makeServiceRequest<AuthResponse>(
      serviceRequest
    );

    text = success ? result : `${translations.moduleToggle.error} ${error}`;
  } else {
    const success = await deactivateModule(userId, moduleId);
    text = success
      ? translations.moduleToggle.success
      : translations.moduleToggle.error;
  }

  return bot.sendMessage(userId, text, await dashboardKeyboardMarkup(userId));
};
